### ISSUE TITLE
Specialize constructors for DOM types that are only exposed to a single global

### ISSUE THREAD

**Issue by jdm** [servo/servo#14071]
Currently, a [WebIDL file](https://github.com/servo/servo/blob/master/components/script/dom/webidls/Range.webidl) with a `[Constructor]` annotation generates Rust code like this:
```rust
unsafe extern fn _constructor(cx: *mut JSContext, argc: u32, vp: *mut JSVal) -> bool {
    return wrap_panic(|| {
        let global = GlobalScope::from_object(JS_CALLEE(cx, vp).to_object());
        let args = CallArgs::from_vp(vp, argc);
        let result: Result<Root<Range>, Error> = Range::Constructor(&global);
```
(where `Range` is the name of the WebIDL interface in question). This means that the `Constructor` method needs to take a `&GlobalScope` argument, so that it can be constructed inside of any JS global environment (eg. a window, a worker, a service worker, etc.).

For interfaces are only exposed to a single global (ie. they have no `[Exposed]` annotation that [explicitly exposes them to multiple globals](https://github.com/servo/servo/blob/7c6793fc28423e518a0ceaae1176048fcb2caa3a/components/script/dom/webidls/TextEncoder.webidl#L6)), we can generate more specialized code instead. We would like to generate:
```rust
unsafe extern fn _constructor(cx: *mut JSContext, argc: u32, vp: *mut JSVal) -> bool {
    return wrap_panic(|| {
        let global = GlobalScope::from_object(JS_CALLEE(cx, vp).to_object());
        let global = global.downcast::<Window>().unwrap();
        let args = CallArgs::from_vp(vp, argc);
        let result: Result<Root<Range>, Error> = Range::Constructor(&global);
```
This will require the `Constructor` method to accept a `&Window` argument instead (and will therefore require updating parts of the code that benefit from this).

The Python code that generates this Rust code lives in `CGClassConstructorHook` in `components/script/dom/bindings/codegen/CodegenRust.py`. We can use `descriptor.interface.exposureSet` to get the list of global names that an interface is exposed in, so we can check if the list only has one entry and downcast to the name present in that list.

There should be no need to run automated tests for these changes; if the resulting code changes build, that's good enough for a PR!

Note: the generated code can be found in `target/debug/build/script-[hash]/out/Bindings/[Interface]Binding.rs`.

**Comment by nox** [comment]
@jdm Can't we allow `let window = Window::from_object(JS_CALLEE(cx, vp).to_object());` directly?

**Comment by jdm** [comment]
That would need to exist first. It seemed easier to start with this.

**Comment by nox** [comment]
Fair enough. :)

**Comment by ajnirp** [comment]
addressed in #14160
